lib/Wikia/src/Persistence/User/Preferences/PreferencePersistence.php

```diff
@@ -228,9 +232,39 @@ private function findWikisWithLocalPreference( ReverseLookupApi $api, $preferenc
 		return $wikiList;
 	}
 
-	private function findUsersWithGlobalPreference( ReverseLookupApi $api, $preferenceName, $value = null ) {
-		$userList = $api->findUsersWithGlobalPreference( $preferenceName, $value );
-		return $userList;
+	/**
+	 * Calls user-preference /reverse-lookup/global/{preferenceName}/users endpoint to get an array of string userIds
+	 * of users with $preferenceName set to $value (or set at all if $value is null)
+	 * Handles paginating API calls for $numUsersToRetrieve > 10000
+	 * @param ReverseLookupApi $api
+	 * @param string $preferenceName
+	 * @param string|null $value
+	 * @param int|null $numUsersToRetrieve
+	 * @param int|null $userIdContinue
+	 * @return array|string[] - userId
+	 * @throws ApiException
+	 */
+	private function findUsersWithGlobalPreference(ReverseLookupApi $api, $preferenceName, $value = null, $numUsersToRetrieve = 1000, $userIdContinue = null ) {
+		if ( $numUsersToRetrieve <= PreferencePersistence::REVERSE_LOOKUP_GLOBAL_USERS_MAX_LIMIT ) {
+			return $api->findUsersWithGlobalPreference( $preferenceName, $value, $numUsersToRetrieve, $userIdContinue );
+		}
+		else {
+			// pagination for API calls
+			$userList = [];
+			$userIdContinue = intval($userIdContinue);
+			while ( count($userList) < $numUsersToRetrieve ){
+				$numUsersInCall = min(PreferencePersistence::REVERSE_LOOKUP_GLOBAL_USERS_MAX_LIMIT, $numUsersToRetrieve - count($userList));
+				$userList = array_merge($userList,
+					$api->findUsersWithGlobalPreference(
+						$preferenceName,
+						$value,
+						$numUsersInCall,
+						$userIdContinue )
+				);
+				$userIdContinue += $numUsersInCall;
```

Comment:
Consider the following scenario (let's assume that the max size of a page is 10, to make it easier to read):

1. A caller retrieves 25 users with ids after 1
2. At the beginning the userList is empty, userIdContinue = 1, numUsersInCall = 10
3. There are no users with ids 8 and 9, so the function returns ids: 1-7, 10-12
4. $userIdContinue += $numUsersInCall makes userIdContinue = 11
5. The code retrieves 10 users with ids after 11.
6. This time all are available and it adds to the list users with ids 11-20
7. The user with id 12 has been added twice to the list.